integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts

```diff
@@ -0,0 +1,1528 @@
+import {
+  ChangeDetectorRef,
+  ElementRef,
+  EventEmitter,
+  Input,
+  Output,
+  Injectable,
+  OnDestroy,
+} from '@angular/core';
+import { BehaviorSubject, Observable, of, Subscription } from 'rxjs';
+import {
+  EpdVisualizationConfig,
+  EpdVisualizationInnerConfig,
+  Ui5Config,
+  VisualizationApiConfig,
+} from '@spartacus/epd-visualization/root';
+import { SceneNodeToProductLookupService } from '@spartacus/epd-visualization/root';
+import { VisualizationLookupService } from '@spartacus/epd-visualization/root';
+
+// @ts-ignore
+import * as ui5 from '@sapui5/ts-types-esm';
+
+import type Core from 'sap/ui/core/Core';
```

Comment:
I'm definitiely missing someting... Can you help me undestand how it's possible we're importing stuff from `sap/*` packages, but they are not declared in our package.json and installed? I can only see `@sapui5/ts-types-esm` in our devDependencies.

Reply:
The import statements are importing type definitions only (UI5 is implemented in JavaScript, not Typescript).
UI5 is being bootstrapped by adding a <script> tag into the DOM that references a script specified by the `bootstrapUrl` configuration item.
There are npm packages for the UI5 JavaScript code (which are separate packages from the type definitions), but it is not straightforward to use them from Angular.
The UI5 tooling uses a custom babel extension to munge code so that these npm packages can be used in TypeScript applications, but I don't think this babel extension is compatible with the Angular build process.
It would be really nice to be able to use the npm packages instead of bootstrapping UI5 from the CDN, but I don't see it happening in what remains of this year.

Reply:
Thanks for the explanation.
1.
> There are npm packages for the UI5 JavaScript code (which are separate packages from the type definitions), but it is not straightforward to use them from Angular. 
Can you help me understand what were the issues?

2.
> The other thing about the UI5 type definitions that causes issues is that the modules have names like sap/ui/core/Core in a package named @sapui5/ts-types-esm.
> 
> It would be really nice to be able to use the npm packages instead of bootstrapping UI5 from the CDN, but I don't see it happening this year.

That makes sense. Btw. I like we lazy-load ui5, three.js and all stuff needed only when EPD is on the page 👍 

3. 
> Note that the parameters in the lambda expression passed to sap.ui.require that provide the constructor functions are hiding variables assigned by import statements earlier in the file.

Just to better understand: do we really need to use `any` types instead of proper types in the arguments? What would happen if we used the proper types for the parameters? (I believe the property names don't have to be named the same as their types - so we can avoid the collision)

Reply:
I think that bootstrapping from UI5 from a URL is actually quite a flexible arrangement.
It is easy to point to a nightly build or a self hosted UI5 build. This has been useful when there has been a need to develop and test using prerelease `sap.ui.vk` code (which has been necessary).
It is also possible to use a floating UI5 version (which has its own set of risks/benefits).

If we do try to use the npm packages then there will be all of this extra code that will be involved in all of the build processes:

| npm package                   | unpacked size |
|-------------------------------|---------------|
| @sapui5/sap.ui.vk             |        7.1 MB |
| @openui5/sap.m                |       11.1 MB |
| @openui5/sap.ui.core          |       26.3 MB |
| @openui5/sap.ui.layout        |        1.1 MB |
| @sapui5/sap.ui.richtexteditor |       25.6 MB |

By using an existing UI5 build we get source maps that work well.
I imagine debugging might be problematic if the UI5 code ends up going through the webpack build process as well.

By using UI5 on the Akamai CDN, the load time is always good regardless of geography.
If we were to use the npm packages, then when the built code is deployed on a geographically distant server the load performance will be significantly worse, which will be a big issue for me and others in my location.

Any given UI5 release has a support period (so customers will/need to move forward with UI5 versions anyway). A floating UI5 version automatically takes care of this.

The main advantage of the NPM package approach is that it is provides complete control over the availability of a specific version of the dependent code (which is a big advantage), but there are a lot of downsides too.

> Can you help me understand what were the issues?

The UI5 NPM packages and related tooling is not oriented towards developing applications in Angular.
The strategy for that is to use UI5 Web Components. However we do not have a web component based viewer at this point.

I had issues with bootstrapping and with the build process. I'm sure it would be possible to make it work with enough effort, but so far I have not been successful and I'm unconvinced that it will actually be best solution.
Even if I could get it to work, there will still be some non-obvious code that needs some explanation required to make it work.  

> do we really need to use any types instead of proper types in the arguments?

Unfortunately that does seem to be the case in this arrangement since there is only a definition in a .d.ts file and no implementation that the Typescript compiler can find, so the compiler says that the expression is not constructable.

Rather than hiding the fields assigned by the import statements I could name the parameters like this which should reduce confusion:

```typescript
      sap.ui.require(
        [
          'sap/ui/vk/ViewManager',
          'sap/ui/vk/Viewport',
          'sap/ui/vk/ViewStateManager',
          'sap/ui/vk/AnimationPlayer',
          'sap/ui/vk/ContentConnector',
          'sap/ui/vk/DrawerToolbar',
        ],
        (
          sap_ui_vk_ViewManager: any,
          sap_ui_vk_Viewport: any,
          sap_ui_vk_ViewStateManager: any,
          sap_ui_vk_AnimationPlayer: any,
          sap_ui_vk_ContentConnector: any,
          sap_ui_vk_DrawerToolbar: any
        ) => {
```

Reply:
Many thanks for your comprehensive explanation! That really helped me understand the context of your decisions.